File: src/main/java/com/ytg123/manhunt/command/HuntersCommand.java
package com.ytg123.manhunt.command;

import com.mojang.brigadier.CommandDispatcher;
import com.mojang.brigadier.context.CommandContext;
import com.mojang.brigadier.exceptions.CommandSyntaxException;
import com.ytg123.manhunt.Manhunt;
import com.ytg123.manhunt.ManhuntUtils;
import net.minecraft.command.argument.EntityArgumentType;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.server.command.ServerCommandSource;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.text.LiteralText;
import net.minecraft.text.TranslatableText;

import java.util.ArrayList;
import java.util.List;

import static net.minecraft.server.command.CommandManager.*;

public class HuntersCommand {
    public static void register(CommandDispatcher<ServerCommandSource> dispatcher) {
        dispatcher.register(
                literal("hunters").requires(source -> source.hasPermissionLevel(2))
                        .then(
                                literal("clear").executes(HuntersCommand::executeClear)
                        )
                        .then(
                                literal("add").then(
                                        argument("target", EntityArgumentType.player()).executes(HuntersCommand::executeAdd)
                                )
                        )
                        .then(
                                literal("get").executes(HuntersCommand::executeGet)
                        )
        );
    }

    private static int executeClear(CommandContext<ServerCommandSource> context) throws CommandSyntaxException {
        boolean playerHasMod = ManhuntUtils.playerHasMod(context);
        ManhuntUtils.hunters.clear();
        if (playerHasMod)
            context.getSource().sendFeedback(new TranslatableText("text.manhunt.command.hunters.clear"), true);
        else
            context.getSource().sendFeedback(new LiteralText("Cleared hunter list!"), false);
        return 1;
    }

    private static int executeAdd(CommandContext<ServerCommandSource> context) throws CommandSyntaxException {
        ServerPlayerEntity target = EntityArgumentType.getPlayer(context, "target");
        boolean playerHasMod = ManhuntUtils.playerHasMod(context);

        // Target is speedrunner
        if (target.equals(ManhuntUtils.speedrunner)) {
            // We check if the source is a player and the player has the mod
            if (playerHasMod) {
                // We send error message using TranslatableText
                context.getSource().sendError(new TranslatableText("text.manhunt.command.hunters.error.speedrunner", target.getDisplayName()));

            // Source is not player or doesn't have mod, sending error message using LiteralText
            } else {
                context.getSource().sendError(new LiteralText("Cannot add ").append(target.getDisplayName()).append(new LiteralText(" as a hunter because they are the speedrunner!")));
            }
            return 1;
        }

        // Check if target is already a hunter
        if (ManhuntUtils.hunters.contains(target)) {
            if (playerHasMod) {
                context.getSource().sendError(new TranslatableText("text.manhunt.command.hunters.error.hunter", target.getDisplayName()));
            } else {
                context.getSource().sendError(new LiteralText("Cannot add ").append(target.getDisplayName()).append(new LiteralText(" as a hunter because they are already a hunter!")));
            }
            return 1;
        }
        ManhuntUtils.hunters.add(target);
        if (playerHasMod)
            context.getSource().sendFeedback(new TranslatableText("text.manhunt.command.hunters.add", target.getDisplayName()), true);
        else
            context.getSource().sendFeedback(new LiteralText("Added ").append(target.getDisplayName()).append(new LiteralText(" to the hunters list!")), true);
        return 1;
    }

    private static int executeGet(CommandContext<ServerCommandSource> context) throws CommandSyntaxException {
        boolean playerHasMod = ManhuntUtils.playerHasMod(context);
        if (ManhuntUtils.hunters.isEmpty()) return 1;
        List<String> hunterNames = new ArrayList<>();
        ManhuntUtils.hunters.forEach(element -> {
            hunterNames.add(element.getDisplayName().asString());
        });
        if (playerHasMod)
            context.getSource().sendFeedback(new TranslatableText("text.manhunt.command.hunters.get", String.join(", ", hunterNames)), false);
        else
            context.getSource().sendFeedback(new LiteralText("Hunters are: " + String.join(", ", hunterNames)), false);
        return 1;
    }
}

File: manhunt-base/src/main/kotlin/io/github/ytg1234/manhunt/command/HuntersCommand.kt
package io.github.ytg1234.manhunt.command

import com.mojang.brigadier.Command
import com.mojang.brigadier.context.CommandContext
import io.github.ytg1234.manhunt.base.CONFIG
import io.github.ytg1234.manhunt.base.fromCmdContext
import io.github.ytg1234.manhunt.base.hunters
import io.github.ytg1234.manhunt.base.playerHasMod
import io.github.ytg1234.manhunt.base.speedRunner
import io.github.ytg1234.manhunt.util.PermedCommand
import io.github.ytg1234.manhunt.util.plus
import io.github.ytg1234.manhunt.util.reset
import mc.aegis.AegisCommandBuilder
import net.minecraft.command.argument.EntityArgumentType
import net.minecraft.item.ItemStack
import net.minecraft.item.Items
import net.minecraft.server.command.CommandManager
import net.minecraft.server.command.ServerCommandSource
import net.minecraft.text.LiteralText
import net.minecraft.text.TranslatableText

/**
 * Used to manage the [hunter list][hunters].
 *
 * @author YTG1234
 */
object HuntersCommand : PermedCommand("hunters", "manhunt.command.hunters", 2) {
    override val cmd: AegisCommandBuilder.() -> AegisCommandBuilder = {
        literal("clear") {
            executes(::executeClear)
        }
        literal("add") {
            custom(CommandManager.argument("target", EntityArgumentType.player())) {
                executes(::executeAdd)
            }
        }
        literal("get") {
            executes(::executeGet)
        }
        this
    }

    /**
     * Clears the hunter list.
     */
    private fun executeClear(context: CommandContext<ServerCommandSource>): Int {
        val playerHasMod: Boolean = playerHasMod(context)
        context.source.minecraftServer.hunters.clear()
        if (playerHasMod) {
            context.source.sendFeedback(TranslatableText("text.manhunt.command.hunters.clear"), true)
        } else {
            context.source.sendFeedback(LiteralText("Cleared hunter list!"), false)
        }
        return Command.SINGLE_SUCCESS
    }

    /**
     * Adds a hunter to the hunter list.
     */
    private fun executeAdd(context: CommandContext<ServerCommandSource>): Int {
        val target = EntityArgumentType.getPlayer(context, "target")
        val playerHasMod: Boolean = playerHasMod(context)

        // Target is speedrunner
        val minecraftServer = context.source.minecraftServer
        if (target.uuid == minecraftServer.speedRunner) {
            // We check if the source is a player and the player has the mod
            if (playerHasMod) {
                // We send error message using TranslatableText
                context.source.sendError(
                    TranslatableText(
                        "text.manhunt.command.hunters.error.speedrunner",
                        target.displayName
                    )
                )

                // Source is not player or doesn't have mod, sending error message using LiteralText
            } else {
                context.source
                    .sendError(
                        LiteralText("Cannot add ").append(target.displayName)
                            .append(LiteralText(" as a hunter because they are the speedrunner!"))
                    )
            }
            return Command.SINGLE_SUCCESS
        }

        // Check if target is already a hunter
        if (minecraftServer.hunters.contains(target.uuid)) {
            if (playerHasMod) {
                context.source.sendError(
                    TranslatableText(
                        "text.manhunt.command.hunters.error.hunter",
                        target.displayName
                    )
                )
            } else {
                context.source.sendError(reset("Cannot add ") + target.displayName + reset(" as a hunter because they are already a hunter!"))
            }
            return Command.SINGLE_SUCCESS
        }
        if (CONFIG.giveCompassWhenSettingHunters) fromCmdContext(context, target.uuid)!!.equip(
            8,
            ItemStack(Items.COMPASS, 1)
        )
        minecraftServer.hunters.add(target.uuid)
        if (playerHasMod) {
            context.source.sendFeedback(TranslatableText("text.manhunt.command.hunters.add", target.displayName), true)
        } else {
            context.source
                .sendFeedback(
                    reset("Added ") + target.displayName + reset(" to the hunters list!"), true
                )
        }
        return Command.SINGLE_SUCCESS
    }

    /**
     * Sends the hunter list.
     */
    private fun executeGet(context: CommandContext<ServerCommandSource>): Int {
        val playerHasMod: Boolean = playerHasMod(context)
        val hunters = context.source.minecraftServer.hunters
        if (hunters.isEmpty()) return 1
        val hunterNames: MutableList<String> = ArrayList()
        hunters.forEach { element ->
            hunterNames.add(
                fromCmdContext(context, element)!!.displayName.asString()
            )
        }
        if (playerHasMod) {
            context.source.sendFeedback(
                TranslatableText(
                    "text.manhunt.command.hunters.get",
                    java.lang.String.join(", ", hunterNames)
                ), false
            )
        } else {
            context.source.sendFeedback(LiteralText("Hunters are: " + java.lang.String.join(", ", hunterNames)), false)
        }
        return Command.SINGLE_SUCCESS
    }
}
